compiler/src/dotty/tools/dotc/transform/SymUtils.scala

```diff
@@ -112,7 +112,14 @@ object SymUtils:
           self.isCoDefinedGiven(res.typeSymbol)
       self.isAllOf(Given | Method) && isCodefined(self.info)
 
-    def useCompanionAsMirror(using Context): Boolean = self.linkedClass.exists && !self.is(Scala2x)
+    def useCompanionAsSumMirror(using Context): Boolean =
+      self.linkedClass.exists
+      && !self.is(Scala2x)
+      && (
+        // self is from source, or companion is a subtype of Sum
```

Comment:
It'd be good to explain why for things which are not from source we need the more precise check (i.e., the difference in code generation between scala 3.0 and 3.1).

Reply:
It's more about the other way round: ideally we can just check is subclass (which is only true in scala 3.1+).  But we call this check before that parent is added, for code in compilation, so have to handle that specially.

Feels like the two new ones would subsume `!self.is(Scala2x)`, right?  (minor)

Reply:
I have clarified the comment to explain why we check if its from source